pil_analyzer/src/evaluator.rs

```diff
@@ -1,104 +1,383 @@
-                self.evaluate(right)?,
+                evaluate_to_number(expr, locals, symbols)?,
             )),
-            Expression::UnaryOperation(op, expr) => {
-                Ok(evaluate_unary_operation(*op, self.evaluate(expr)?))
+            Expression::LambdaExpression(lambda) => {
+                // TODO only copy the part of the environment that is actually referenced?
+                (Closure {
+                    parameter_count: lambda.params.len(),
+                    body: lambda.body.as_ref(),
+                    environment: locals.to_vec(),
+                })
+                .into()
             }
-            Expression::LambdaExpression(_) => {
-                Err("Cannot evaluate lambda expression.".to_string())
+            Expression::IndexAccess(index_access) => {
+                match evaluate(&index_access.array, locals, symbols)? {
+                    Value::Array(elements) => {
+                        let index = evaluate_to_number(&index_access.index, locals, symbols)?;
+                        // TODO avoid conversion to bigint
+                        if index.to_arbitrary_integer() > elements.len().into() {
+                            Err(format!("Index access out of bounds: Tried to access element {index} of array of size {}.", elements.len()))?;
+                        }
+                        elements
+                            .into_iter()
+                            .nth(index.to_degree() as usize)
+                            .unwrap()
+                    }
+                    e => Err(format!("Expected array, but got {e}"))?,
+                }
             }
-            Expression::IndexAccess(_) => Err("Cannot evaluate index access.".to_string()),
             Expression::FunctionCall(FunctionCall { id, arguments }) => {
-                let arg_values = arguments
+                let function = evaluate_reference(id, locals, symbols)?;
+                let arguments = arguments
                     .iter()
-                    .map(|a| self.evaluate(a))
+                    .map(|a| evaluate(a, locals, symbols).map(Rc::new))
                     .collect::<Result<Vec<_>, _>>()?;
-                assert!(arg_values.len() == 1);
-                let values = &self.function_cache[id.to_string().as_str()];
-                Ok(values[arg_values[0].to_degree() as usize % values.len()])
+                match function {
+                    Value::Closure(Closure {
+                        parameter_count,
+                        body,
+                        environment,
+                    }) => {
+                        assert_eq!(parameter_count, arguments.len());
+
+                        let local_vars =
+                            arguments.into_iter().chain(environment).collect::<Vec<_>>();
+
+                        evaluate(body, &local_vars, symbols)?
+                    }
+                    Value::Custom(values) => values.eval_function_application(&arguments)?,
```

Comment:
It might make sense to pass `symbols` here as a function argument. In implementations it turned out that we needed to store a reference inside the custom value. We could avoid that by just passing `symbols` here.